### PR TITLE
feat(compiler-cli): add extended diagnostic to warn for missing aster…

### LINKED ISSUE #50469: Structural Directives should be compile errors if the * is missing
### Which @angular/* package(s) are relevant/related to the feature request?

compiler

### Description

Recently, had a 30 minute delay in productivity because of a missing `*` in `*ngIf`.

There was an error message thrown "No Provider for TemplateRef" - Which is fantastic, I get a line number pointing me to the template; but this should be a compile-time error.

Example:
````html
<div *ngIf="thisWorks">
    I work! But the entire TemplateRef will fail due to a syntax error that isn't being reported.
    <div ngIf="thisBreaks">All it takes is one errant keystroke and this will break at runtime.</div>
</div>
````

This is a hairy problem, because for a while, I centered my attention around "thisWorks". I was trying to figure out, is it that I have a ChangeDetectorRef I need to `markForCheck`, is it an issue with the logic powering thisWorks.... It was only after realizing that the TemplateRef error was connected to "thisBreaks", that I recognized my mistake. It didn't tell me that I was missing a *, just that a TemplateRef could not be found.

Having a more clear error message - or ideally, a red squiggly for ngIf's missing *, would be helpful.

### Proposed solution

Make it possible for directives which inject a TemplateRef to optionally require use of the `*` syntax. This should be a compiler error.

I recognize that there may be legitimate circumstances where a TemplateRef is injected on a directive, but you do not expect or require the `*` syntax. However, for things like *ngIf this should be a compiler error.

### Alternatives considered

I could just always type with perfect accuracy. But I've tried that, and it doesn't always work.

### PR DESCRIPTION
**Pull request by naaajii**
…isk on structural directives

if user forgets to add asterisk before ngFor, ngIf, ngSwitchCase or ngSwitchDefault. this diagnostic will warn user about missing asterisk before structural directives.

fixes #50469

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

We don't get a warning for missing `*` before ngFor, ngIf, ngSwitchCase or ngSwitchDefault.

Issue Number: #50469

## What is the new behavior?
We get a warning for missing `*` before ngFor, ngIf, ngSwitchCase or ngSwitchDefault.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
this is my attempt to add an extended diagnostic for missing asterisks before structural directives. it might not be perfect, looking forward to improve!

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  override visitNode(
+      ctx: TemplateContext<ErrorCode.MISSING_ASTERISK_STRUCTURAL_DIRECTIVES>,
+      component: ts.ClassDeclaration, node: TmplAstNode|AST):
+      NgTemplateDiagnostic<ErrorCode.MISSING_ASTERISK_STRUCTURAL_DIRECTIVES>[] {
+    if (!(node instanceof TmplAstElement)) return [];
+
+    if (node.attributes.length === 0) return [];
+
+    const directives: string[] = ['ngFor', 'ngIf', 'ngSwitchCase', 'ngSwitchDefault'];
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts:33]: nit: you could define this on outside the class.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        node.attributes.findIndex((attribute) => directives.includes(attribute.name));
+
+    if (attrIndex === -1) {
+      return [];
+    }
+
+    const errorString = 'You seems to be missing `*` for ' + node.attributes[attrIndex].name + '.';
+    const diagnostic = ctx.makeTemplateDiagnostic(node.sourceSpan, errorString);
+    return [diagnostic];
+  }
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts:45]: Could be simplified a bit with `find()`
```suggestion
    const attr =   node.attributes.find((attribute) => directives.includes(attribute.name));

    if (!attr) {
      return [];
    }

    const errorString = 'You seems to be missing `*` for ' + attr.name + '.';
    const diagnostic = ctx.makeTemplateDiagnostic(node.sourceSpan, errorString);
    return [diagnostic];
  }
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    const directives: string[] = ['ngFor', 'ngIf', 'ngSwitchCase', 'ngSwitchDefault'];
+
+    const attrIndex: number =
+        node.attributes.findIndex((attribute) => directives.includes(attribute.name));
+
+    if (attrIndex === -1) {
+      return [];
+    }
+
+    const errorString = 'You seems to be missing `*` for ' + node.attributes[attrIndex].name + '.';
```
Comment (JeanMeche) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts:42]: What do you think of something like ``` `You are missing the asterisk `*` from the structural directive ${....}.` ``` ?

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts (hunk 1 above)
Comment (Naaajii) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts]: I was actually gonna do this in the initial commit but I missed it later on! :sweat_smile:

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts (hunk 3 above)
Comment (Naaajii) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts]: sounds more valid than my initial error message!

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts (hunk 2 above)
Comment (Naaajii) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_asterisk_structural_directives/index.ts]: sounds good to me & resolved!

**Comment by Naaajii** [comment]
@JoostK thoughts on this? I don't want to sound desperate or anything, but I would love to have this PR as my sixteen on my pull shark badge. :smile:

**Pull request closed without merging**